### PR TITLE
Benchmarks for find_entry

### PR DESCRIPTION
**Pull request by arkpar**
```
$ cargo +nightly bench --features bench bench
    Finished bench [optimized + debuginfo] target(s) in 0.06s
     Running unittests src/lib.rs (target/release/deps/parity_db-21d11d42e7884130)

running 2 tests
test index::test::bench_find_entry     ... bench:          29 ns/iter (+/- 0)
test index::test::bench_find_entry_sse ... bench:          13 ns/iter (+/- 0)
```

Also added a small optimization for `chunk_at`

### CODE REVIEW

File: src/index.rs [hunk 1]
```diff
@@ -287,7 +305,12 @@ impl IndexTable {
...
 
 	#[cfg(any(not(target_feature = "sse2"), test))]
-	fn find_entry_base(&self, key_prefix: u64, sub_index: usize, chunk: &[u8]) -> (Entry, usize) {
+	fn find_entry_base(
+		&self,
+		key_prefix: u64,
+		sub_index: usize,
+		chunk: &[u8; CHUNK_LEN],
+	) -> (Entry, usize) {
 		assert!(chunk.len() >= CHUNK_ENTRIES * 8);
```
Comment (cheme) [inline src/index.rs:314]: Is the assert still needed ?

File: src/index.rs [hunk 2]
```diff
@@ -641,4 +670,46 @@ mod test {
...
+			IndexTable::write_entry(&e, i, &mut chunk);
+		}
+
+		let mut index = 0;
+
+		b.iter(|| {
+			let i = index % 64;
+			let x = f(&table, keys[i], 0, &chunk).1;
+			assert_eq!(x, i);
+			index += 1;
```
Comment (cheme) [inline src/index.rs:699]: Would make sense to me to have the 64 query under iter (so nb iter don't interfer (actually with this use case it probably does not really matter)).

File: src/index.rs [hunk 3]
```diff
@@ -641,4 +670,46 @@ mod test {
...
+		F: Fn(&IndexTable, u64, usize, &[u8; CHUNK_LEN]) -> (Entry, usize),
+	>(
+		b: &mut Bencher,
+		f: F,
+	) {
+		let table =
+			IndexTable { id: TableId(18), map: RwLock::new(None), path: Default::default() };
+		let mut chunk = [0u8; 512];
+		let mut keys = [0u64; 64];
+		let mut rng = rand::thread_rng();
```
Comment (cheme) [inline src/index.rs:685]: Not too sure if we should bench on random content (could be just random from a fix seed).

**Review by cheme — APPROVED** [review]

**Review by Tpt — APPROVED** [review]
+1 to @cheme suggestions

File: src/index.rs (hunk 2 above)
Comment (arkpar) [reply src/index.rs:699]: Under real workload chunks are never full. I'd guess average occupancy is about 33%. With just a few entries base variant will probably be faster, while with 64 entries simd will probably be faster.  This cycle averages it.

File: src/index.rs (hunk 3 above)
Comment (arkpar) [reply src/index.rs]: OK, even though It does not matter too much what the entry bits are, as long as they are different.

**Pull request merged**